### PR TITLE
📝 Fix comments on RLP encoding scheme

### PR DESCRIPTION
**Pull request by pcaversaccio**
Fixes https://github.com/transmissions11/solmate/issues/207#issuecomment-1205862986.

### CODE REVIEW

**Comment by pcaversaccio** [comment]
if you want we can also go up to `uint64`. I implemented it here: https://github.com/pcaversaccio/create-util/blob/main/contracts/Create.sol#L98

File: src/utils/LibRLP.sol [hunk 1]
```diff
@@ -10,21 +10,27 @@ library LibRLP {
...
+        // 0x84 = 0x80 + 0x04 (= the bytes length of the nonce, 4 bytes, in hex).
         if (nonce <= type(uint8).max)  return keccak256(abi.encodePacked(bytes1(0xd7), bytes1(0x94), deployer, bytes1(0x81), uint8(nonce))).fromLast20Bytes();
         if (nonce <= type(uint16).max) return keccak256(abi.encodePacked(bytes1(0xd8), bytes1(0x94), deployer, bytes1(0x82), uint16(nonce))).fromLast20Bytes();
         if (nonce <= type(uint24).max) return keccak256(abi.encodePacked(bytes1(0xd9), bytes1(0x94), deployer, bytes1(0x83), uint24(nonce))).fromLast20Bytes();
 
-        // More details about RLP encoding can be found here: https://eth.wiki/fundamentals/rlp
-        // 0xda = 0xc0 (short RLP prefix) + 0x16 (length of: 0x94 ++ proxy ++ 0x84 ++ nonce)
-        // 0x94 = 0x80 + 0x14 (0x14 = the length of an address, 20 bytes, in hex)
-        // 0x84 = 0x80 + 0x04 (0x04 = the bytes length of the nonce, 4 bytes, in hex)
-        // We assume nobody can have a nonce large enough to require more than 32 bytes.
```
Comment (transmissions11) [inline src/utils/LibRLP.sol:27]: oh shoot this should have said 32 bits

**Comment by transmissions11** [comment]
> if you want we can also go up to `uint64`. I implemented it here: [pcaversaccio/create-util@`main`/contracts/Create.sol#L98](https://github.com/pcaversaccio/create-util/blob/main/contracts/Create.sol?rgh-link-date=2022-08-05T09%3A53%3A06Z#L98)

you'd have to send ~117670 tx/s a day for 100 years to need a 64 bit nonce? dont think thats worth it

File: src/utils/LibRLP.sol (hunk 1 above)
Comment (pcaversaccio) [reply src/utils/LibRLP.sol:27]: Yes, I corrected this in line 15: "However, we assume nobody can have a nonce large enough to require more than 4 bytes."

**Comment by pcaversaccio** [comment]
> > if you want we can also go up to `uint64`. I implemented it here: [pcaversaccio/create-util@`main`/contracts/Create.sol#L98](https://github.com/pcaversaccio/create-util/blob/main/contracts/Create.sol?rgh-link-date=2022-08-05T09%3A53%3A06Z#L98)
> 
> you'd have to send ~117670 tx/s a day for 100 years to need a 64 bit nonce? dont think thats worth it

that contract may be around for a long time - we could add a simple check instead (I prefer custom errors tho ;-)) : 
```solidity
require(nonce <= type(uint32).max, "RLP: invalid nonce value);
```

**Comment by pcaversaccio** [comment]
I just realised - why are there no tests on RLP? For my 🐍 snekmate Vyper contracts I actually implemented tests for that [here](https://github.com/pcaversaccio/snekmate/blob/main/test/utils/CreateAddress.t.sol) (will refactor them a bit).

**Comment by pcaversaccio** [comment]
@transmissions11 since the CTF is over it's time to reconsider this PR looool

**Comment by pcaversaccio** [comment]
@transmissions11 
image

**Comment by pcaversaccio** [comment]
Ok, another follow-up - wen merge @transmissions11?

**Comment by pcaversaccio** [comment]
image

File: src/utils/LibRLP.sol [hunk 2]
```diff
@@ -10,21 +10,27 @@ library LibRLP {
 
     // prettier-ignore
     function computeAddress(address deployer, uint256 nonce) internal pure returns (address) {
+    
+        // @dev The theoretical allowed limit, based on EIP-2681, for an account nonce is 2**64-2: https://eips.ethereum.org/EIPS/eip-2681.
```
Comment (transmissions11) [inline src/utils/LibRLP.sol:14]:
```suggestion
        // The theoretical allowed limit, based on EIP-2681, for an account nonce is 2**64-2: https://eips.ethereum.org/EIPS/eip-2681.
```

File: src/utils/LibRLP.sol [hunk 3]
```diff
@@ -10,21 +10,27 @@ library LibRLP {
...
+
         // The integer zero is treated as an empty byte string, and as a result it only has a length prefix, 0x80, computed via 0x80 + 0.
-        // A one byte integer uses its own value as its length prefix, there is no additional "0x80 + length" prefix that comes before it.
         if (nonce == 0x00)             return keccak256(abi.encodePacked(bytes1(0xd6), bytes1(0x94), deployer, bytes1(0x80))).fromLast20Bytes();
+        
+        // A one-byte integer in the [0x00, 0x7f] range uses its own value as a length prefix, there is no additional "0x80 + length" prefix that precedes it.
         if (nonce <= 0x7f)             return keccak256(abi.encodePacked(bytes1(0xd6), bytes1(0x94), deployer, uint8(nonce))).fromLast20Bytes();
 
-        // Nonces greater than 1 byte all follow a consistent encoding scheme, where each value is preceded by a prefix of 0x80 + length.
+        // In the case of `nonce > 0x7f` and `nonce <= type(uint8).max`, we have the following
```
Comment (transmissions11) [inline src/utils/LibRLP.sol:24]:
```suggestion
        // In the case of nonce > 0x7f and nonce <= type(uint8).max, we have the following
```

File: src/utils/LibRLP.sol [hunk 4]
```diff
@@ -10,21 +10,27 @@ library LibRLP {
...
         if (nonce <= type(uint8).max)  return keccak256(abi.encodePacked(bytes1(0xd7), bytes1(0x94), deployer, bytes1(0x81), uint8(nonce))).fromLast20Bytes();
         if (nonce <= type(uint16).max) return keccak256(abi.encodePacked(bytes1(0xd8), bytes1(0x94), deployer, bytes1(0x82), uint16(nonce))).fromLast20Bytes();
         if (nonce <= type(uint24).max) return keccak256(abi.encodePacked(bytes1(0xd9), bytes1(0x94), deployer, bytes1(0x83), uint24(nonce))).fromLast20Bytes();
 
-        // More details about RLP encoding can be found here: https://eth.wiki/fundamentals/rlp
-        // 0xda = 0xc0 (short RLP prefix) + 0x16 (length of: 0x94 ++ proxy ++ 0x84 ++ nonce)
-        // 0x94 = 0x80 + 0x14 (0x14 = the length of an address, 20 bytes, in hex)
-        // 0x84 = 0x80 + 0x04 (0x04 = the bytes length of the nonce, 4 bytes, in hex)
-        // We assume nobody can have a nonce large enough to require more than 32 bytes.
+        // Case for `nonce > uint24` and `nonce <= type(uint32).max`.
```
Comment (transmissions11) [inline src/utils/LibRLP.sol:33]:
```suggestion
        // Case for nonce > uint24 and nonce <= type(uint32).max.
```

File: src/utils/LibRLP.sol [hunk 5]
```diff
@@ -10,21 +10,27 @@ library LibRLP {
 
     // prettier-ignore
     function computeAddress(address deployer, uint256 nonce) internal pure returns (address) {
+    
+        // The theoretical allowed limit, based on EIP-2681, for an account nonce is 2**64-2: https://eips.ethereum.org/EIPS/eip-2681.
```
Comment (transmissions11) [inline src/utils/LibRLP.sol:14]:
```suggestion
    function computeAddress(address deployer, uint256 nonce) internal pure returns (address) {
        // The theoretical allowed limit, based on EIP-2681, for an account nonce is 2**64-2: https://eips.ethereum.org/EIPS/eip-2681.
```

**Comment by transmissions11** [comment]
k this lgtm, thanks for your patience

**Comment by pcaversaccio** [comment]
:-D thanks for the review and merge!

**Pull request merged**